From 8f6db7a2663ab2cfb50569b1ea74e02395703245 Mon Sep 17 00:00:00 2001
From: Rudi <r.ihle@s-t.de>
Date: Sat, 26 Dec 2015 19:15:55 +0100
Subject: [PATCH] video: mxc_hdmi: Simplify aspect ratio setup for AVI
 infoframe

Either use the aspect ratio flags already present in fb_videomode
or calulate them from current resolution.

Signed-off-by: Rudi <r.ihle@s-t.de>
---
 drivers/video/mxc/mxc_hdmi.c | 18 +++++++-----------
 1 file changed, 7 insertions(+), 11 deletions(-)

diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index 170704d..e5d6bc5 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -1418,21 +1418,17 @@ static void hdmi_config_AVI(struct mxc_hdmi *hdmi)
 	u8 under_scan;
 	u8 act_ratio, coded_ratio, colorimetry, ext_colorimetry;
 	struct fb_videomode mode;
-	const struct fb_videomode *edid_mode;
-	bool aspect_16_9;
 
 	dev_dbg(&hdmi->pdev->dev, "set up AVI frame\n");
 
 	fb_var_to_videomode(&mode, &hdmi->fbi->var);
-	/* Use mode from list extracted from EDID to get aspect ratio */
-	if (!list_empty(&hdmi->fbi->modelist)) {
-		edid_mode = fb_find_nearest_mode(&mode, &hdmi->fbi->modelist);
-		if (edid_mode->vmode & FB_VMODE_ASPECT_16_9)
-			aspect_16_9 = true;
+
+	if (!(mode.vmode & FB_VMODE_ASPECT_MASK)) {
+		if( mode.xres <= (mode.yres / 3) * 4)
+			mode.vmode |= FB_VMODE_ASPECT_4_3;
 		else
-			aspect_16_9 = false;
-	} else
-		aspect_16_9 = false;
+			mode.vmode |= FB_VMODE_ASPECT_16_9;
+	}
 
 	/********************************************
 	 * AVI Data Byte 1
@@ -1464,7 +1460,7 @@ static void hdmi_config_AVI(struct mxc_hdmi *hdmi)
 	 ********************************************/
 
 	/*  Set the Aspect Ratio */
-	if (aspect_16_9) {
+	if (mode.vmode & FB_VMODE_ASPECT_16_9) {
 		act_ratio = HDMI_FC_AVICONF1_ACTIVE_ASPECT_RATIO_16_9;
 		coded_ratio = HDMI_FC_AVICONF1_CODED_ASPECT_RATIO_16_9;
 	} else {
-- 
1.9.1

